### PR TITLE
🖨️ Feat/1143 Exception Manifest TSV

### PR DESCRIPTION
**Pull request by demariadaniel**
## Link to Issue

#1143 

## Description

- Adds `exception_manifest.tsv` to Clinical Data TSV download
- Updates manifest function to retrieve all Donors when arguments are empty []
- TS Updates based on feedback from https://github.com/icgc-argo/argo-clinical/pull/1153#pullrequestreview-1923753630
- Breaks down initial unit test into several smaller test cases

### Type of Change

- [ ] Bug
- [ ] Refactor
- [x] New Feature
- [ ] Release Candidate

### Checklist before requesting review:

- [x] Check branch (code change PRs go to `develop` not master)
- [x] Check copyrights for new files
- [x] Manual testing
- [x] Regression tests completed and passing (double check number of tests).
- [x] Spelling has been checked.
- [x] Updated swagger docs accordingly (check it's still valid)
- [ ] Set `validationDependency` in meta tag for [Argo Dictionary](https://github.com/icgc-argo/argo-dictionary) fields used in code

### CODE REVIEW

File: src/submission/exceptions/exceptions.ts [hunk 1]
```diff
@@ -287,8 +291,11 @@ export async function getExceptionManifestRecords(
 	filters: { donorIds: number[]; submitterDonorIds: string[] },
 ): Promise<ExceptionManifestRecord[]> {
 	const { donorIds, submitterDonorIds: querySubmitterIds } = filters;
+	const filteredDonors = Boolean(donorIds.length || querySubmitterIds.length);
 
-	const donorsByDonorId = await getDonorsByIds(donorIds);
+	const donorsByDonorId = filteredDonors
+		? await getDonorsByIds(donorIds)
+		: (await getDonors(programId)).map((donor) => donor);
```
Comment (demariadaniel) [inline src/submission/exceptions/exceptions.ts:298]: If donorIds and submitterIds are empty, then we return all exceptions for all donors

File: src/clinical/api/clinical-api.ts [hunk 2]
```diff
@@ -55,12 +57,26 @@ export const parseDonorIdList = (donorIds: string) =>
...
 	const zip = new AdmZip();
 	data.forEach((entityData) => {
 		const tsvData = TsvUtils.convertJsonRecordsToTsv(entityData.records, entityData.entityFields);
 		zip.addFile(`${entityData.entityName}.tsv`, Buffer.alloc(tsvData.length, tsvData));
 	});
+	if (exceptionManifest && exceptionManifest.exceptions.length) {
+		const headers = exceptionManifest.exceptions
+			.map((exception) => Object.keys(exception))
+			.flat()
+			.filter((key, index, keyArray) => keyArray.indexOf(key) === index);
```
Comment (ciaranschutte) [inline src/clinical/api/clinical-api.ts:76]: 100% valid code. please don't change it. just sharing an alternative 

```ts
const headers = new Set();
exceptionManifest.exceptions.forEach(exception => Object.keys(exception).foreach(key => headers.add(key))
```

File: src/submission/exceptions/exceptions.ts (hunk 1 above)
Comment (ciaranschutte) [reply src/submission/exceptions/exceptions.ts:298]:
```ts
[...].map((x) => x)
```
will just return the original array, nothing is changing about the property you're iterating on

**Review by ciaranschutte — DISMISSED** [review]
an Array.map not needed but apart from that looks ok 👍

**Review by ciaranschutte — APPROVED** [review]

**Pull request merged**